crates/boojum/src/gadgets/tower_extension/fq6.rs

```diff
@@ -0,0 +1,882 @@
+
+        Self::new(c0, c1, c2)
+    }
+
+    /// Multiplies the element `a=a0+a1*v+a2*v^2` in `Fq6` by the element `b = b1*v`
+    pub fn mul_by_c1<CS>(&mut self, cs: &mut CS, c1: &mut Fq2<F, T, NN, P::Ex2>) -> Self
+    where
+        CS: ConstraintSystem<F>,
+    {
+        let mut b_b = self.c1.mul(cs, c1);
+        let mut tmp = self.c1.add(cs, &mut self.c2);
+
+        let mut t1 = c1.mul(cs, &mut tmp);
+        let mut t1 = t1.sub(cs, &mut b_b);
+        let t1 = t1.mul_by_nonresidue(cs);
+
+        let mut tmp = self.c0.add(cs, &mut self.c1);
+        let mut t2 = c1.mul(cs, &mut tmp);
+        let t2 = t2.sub(cs, &mut b_b);
+
+        Self::new(t1, t2, b_b)
+    }
+
+    /// Multiplies the element `a=a0+a1*v+a2*v^2` in `Fq6` by the element in `NonNativeField`
+    pub fn mul_by_fq<CS>(&mut self, cs: &mut CS, c0: &mut NN) -> Self
+    where
+        CS: ConstraintSystem<F>,
+    {
+        // Simply multiply element-wise
+        let t0 = self.c0.mul_c0(cs, c0);
+        let t1 = self.c1.mul_c0(cs, c0);
+        let t2 = self.c2.mul_c0(cs, c0);
+
+        Self::new(t0, t1, t2)
+    }
+
+    /// Multiplies the element `a=a0+a1*v+a2*v^2` in `Fq6` by the element `c=c0` in `Fq2`
+    pub fn mul_by_c0<CS>(&mut self, cs: &mut CS, c0: &mut Fq2<F, T, NN, P::Ex2>) -> Self
+    where
+        CS: ConstraintSystem<F>,
+    {
+        // Simply multiply element-wise
+        let t0 = self.c0.mul(cs, c0);
+        let t1 = self.c1.mul(cs, c0);
+        let t2 = self.c2.mul(cs, c0);
+
+        Self::new(t0, t1, t2)
+    }
+
+    /// Multiplies the element `a=a0+a1*v+a2*v^2` in `Fq6` by the element `c2*v^2`
+    pub fn mul_by_c2<CS>(&mut self, cs: &mut CS, c2: &mut Fq2<F, T, NN, P::Ex2>) -> Self
+    where
+        CS: ConstraintSystem<F>,
+    {
+        // Suppose a = a0 + a1*v + a2*v^2. In this case,
+        // (a0 + a1*v + a2*v^2) * c2 * v^2 =
+        // a1*c2*\xi + a2*c2*\xi*v + a0*c2*v^2
+        // NOTE: There might be a better way to calculate three coefficients
+        // without using 3 multiplications and 2 mul_by_nonresidues, similarly to mul_by_c1
+
+        // Setting coefficients
+        let mut a0 = self.c0.clone();
+        let mut a1 = self.c1.clone();
+        let mut a2 = self.c2.clone();
+
+        // new_c0 <- a1*c2*\xi
+        let mut new_c0 = a1.mul(cs, c2);
+        new_c0 = new_c0.mul_by_nonresidue(cs);
+
+        // new_c1 <- a2*c2*\xi
+        let mut new_c1 = a2.mul(cs, c2);
+        new_c1 = new_c1.mul_by_nonresidue(cs);
+
+        // new_c2 <- a0*c2
+        let new_c2 = a0.mul(cs, c2);
+
+        Self::new(new_c0, new_c1, new_c2)
+    }
+
+    /// Multiplies the element `a=a0+a1*v+a2*v^2` in `Fq6` by the element `b = b0+b1*v`
+    pub fn mul_by_c0c1<CS>(
+        &mut self,
+        cs: &mut CS,
+        c0: &mut Fq2<F, T, NN, P::Ex2>,
+        c1: &mut Fq2<F, T, NN, P::Ex2>,
+    ) -> Self
+    where
+        CS: ConstraintSystem<F>,
+    {
+        let mut a_a = self.c0.mul(cs, c0);
```

Comment:
(a0+a1*v+a2*v^2)(b0+b1*v) 
a0b0 +a1b0v + a2b0v^2 + a0b1v + a1b1v^2 + a2b1v^3
c0 = a0b0 + a2b1 xi   =>  3m + 1a
c1 = a1b0 + a0b1   => (a1 + a0)(b0 + b1) - a1b1 - a0b0 => 4A + 2M
c2 = a2b0 + a1b1   => 1m + 1A
6m + 6a 

c0 = a0b0 + a2b1 xi   =>  3m + 1a
c1 = a1b0 + a0b1   => (a1 + a0)(b0 + b1) - a1b1 - a0b0 => 4A + 2M
c2 = a2b0 + a1b1   => (a2 + a1)(b0 + b1) - a2b1 - a1b1 => 3A + 1M
 6M + 8A


in your optimization this is 6M + 10A

Reply:
Indeed, thanks